queue-manager/rango-preset/src/actions/common/produceNextStateForTransaction.ts

```diff
@@ -0,0 +1,167 @@
+): Result<NextTransactionState, NextTransactionStateError> {
+  const { getStorage, context } = actions;
+  const { isMobileWallet } = context;
+  const swap = getStorage().swapDetails;
+
+  const currentStep = getCurrentStep(swap)!;
+
+  const sourceWallet = getRelatedWallet(swap, currentStep);
+  const mobileWallet = isMobileWallet(sourceWallet?.walletType);
+  const tx = getCurrentStepTx(currentStep);
+  const isApproval = isApprovalCurrentStepTx(currentStep);
+
+  if (!tx || !tx.type) {
+    return new Err({
+      nextStatus: 'failed',
+      nextStepStatus: 'failed',
+      message: 'Unexpected Error: tx is null!',
+      details: undefined,
+      errorCode: 'CLIENT_UNEXPECTED_BEHAVIOUR',
+    });
+  }
+
+  const hasAlreadyProceededToSign =
+    typeof swap.hasAlreadyProceededToSign === 'boolean';
+
+  if (isApproval) {
+    return new Ok({
+      nextStatus: undefined,
+      nextStepStatus: 'waitingForApproval',
+      details:
+        'Waiting for approve transaction to be mined and confirmed successfully',
+      message: `Waiting for approval of ${currentStep?.fromSymbol} coin ${
+        mobileWallet ? 'on your mobile phone!' : ''
+      }`,
+    });
+  } else if (hasAlreadyProceededToSign) {
+    return new Err({
+      message: 'Transaction is expired. Please try again.',
+      nextStepStatus: 'failed',
+      nextStatus: 'failed',
+      details: '',
+      errorCode: 'TX_EXPIRED',
+    });
+  }
+
+  return new Ok({
+    message: 'Executing transaction ...',
+    nextStepStatus: 'running',
+    nextStatus: 'running',
+    details: `${mobileWallet ? 'Check your mobile phone!' : ''}`,
+  });
+}
+
+export function onNextStateError(
+  actions: ExecuterActions<SwapStorage, SwapActionTypes, SwapQueueContext>,
+  error: NextTransactionStateError
+) {
+  const { getStorage, setStorage } = actions;
+  const swap = getStorage().swapDetails;
+  const currentStep = getCurrentStep(swap)!;
+  const isApproval = isApprovalCurrentStepTx(currentStep);
+
+  const hasAlreadyProceededToSign =
+    typeof swap.hasAlreadyProceededToSign === 'boolean';
+
+  const updatedResult = updateSwapStatus({
+    getStorage,
+    setStorage,
+
+    nextStepStatus: error.nextStepStatus,
+    nextStatus: error.nextStatus,
+    message: error.message,
+    details: error.details,
+
+    hasAlreadyProceededToSign: isApproval
+      ? undefined
+      : hasAlreadyProceededToSign,
+    errorCode: hasAlreadyProceededToSign ? 'TX_EXPIRED' : undefined,
+  });
+
+  notifier({
+    event: {
+      type: StepEventType.FAILED,
+      reason: error.message,
+      reasonCode: updatedResult.failureType ?? DEFAULT_ERROR_CODE,
+      inputAmount: getLastFinishedStepInput(swap),
+      inputAmountUsd: getLastFinishedStepInputUsd(swap),
+    },
+    ...updatedResult,
+  });
+}
+
+export function onNextStateOk(
+  actions: ExecuterActions<SwapStorage, SwapActionTypes, SwapQueueContext>,
+  ok: NextTransactionState
```

Comment:
I don't think the name of this parameter effectively communicates its meaning.

```suggestion
  state: NextTransactionState
```

Reply:
Err and Ok are types from `ts-result`, keep it this way is more explicit I guess. we did same for `onNextStateError` where the second param named `error`.